### PR TITLE
feat: github support scope and transformation rule

### PR DESCRIPTION
**Pull request by mindlesscloud**
### Summary
1. add table `_tool_github_transformation_rules`
2. add `transformation_rule_id`  to `_tool_github_repos`
3. remove default transformation rules
4. supply HTTP endpoints for the entities `transformation rules` and `GitHub repos`

### Does this close any open issues?
No, partially fulfilled issue #3468

### Screenshots
image
image

### CODE REVIEW

File: plugins/jira/models/board.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ type JiraBoard struct {
 	common.NoPKModel
 	ConnectionId         uint64 `gorm:"primaryKey"`
 	BoardId              uint64 `gorm:"primaryKey"`
-	ScopeId              string
+	ScopeId              string `gorm:"type:varchar(255)"`
```
Comment (warren830) [inline plugins/jira/models/board.go:28]: is this scopeId a domain id?

File: plugins/github/api/scope.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+// GetScopeList get Github repos
+// @Summary get Github repos
+// @Description get Github repos
+// @Tags plugins/github
+// @Param connectionId path int false "connection ID"
+// @Success 200  {object} []models.GithubRepo
+// @Failure 400  {object} shared.ApiBody "Bad Request"
+// @Failure 500  {object} shared.ApiBody "Internal Error"
+// @Router /plugins/github/connections/{connectionId}/scopes/ [GET]
+func GetScopeList(input *core.ApiResourceInput) (*core.ApiResourceOutput, errors.Error) {
```
Comment (klesh) [inline plugins/github/api/scope.go:111]: the number of total scopes could be huge over time. we should add pagination to this API.

File: plugins/github/api/scope.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+func extractParam(params map[string]string) (uint64, uint64) {
+	connectionId, _ := strconv.ParseUint(params["connectionId"], 10, 64)
+	repoId, _ := strconv.ParseUint(params["repoId"], 10, 64)
+	return connectionId, repoId
+}
+
+func verifyRepo(repo *models.GithubRepo) errors.Error {
+	if repo.ConnectionId == 0 {
```
Comment (klesh) [inline plugins/github/api/scope.go:154]: Could it be negative?

File: plugins/github/api/scope.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+		return errors.BadInput.New("invalid connectionId")
+	}
+	if repo.GithubId == 0 {
```
Comment (klesh) [inline plugins/github/api/scope.go:157]: same as above

File: plugins/github/api/scope.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+		return errors.BadInput.New("invalid repoId")
+	}
+	if repo.ScopeId != strconv.Itoa(repo.GithubId) {
```
Comment (klesh) [inline plugins/github/api/scope.go:160]: same as above

File: plugins/github/api/transformation_rule.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+// GetTransformationRuleList return all transformation rules
+// @Summary return all transformation rules
+// @Description return all transformation rules
+// @Tags plugins/github
+// @Success 200  {object} []models.TransformationRules
+// @Failure 400  {object} shared.ApiBody "Bad Request"
+// @Failure 500  {object} shared.ApiBody "Internal Error"
+// @Router /plugins/github/transformation_rules [GET]
+func GetTransformationRuleList(input *core.ApiResourceInput) (*core.ApiResourceOutput, errors.Error) {
```
Comment (klesh) [inline plugins/github/api/transformation_rule.go:118]: pagination

File: plugins/github/tasks/task_data.go [hunk 7]
```diff
@@ -55,40 +56,43 @@ func DecodeAndValidateTaskOptions(options map[string]interface{}) (*GithubOption
...
-	}
-	if op.DeploymentPattern == "" {
-		op.DeploymentPattern = "(?i)deploy"
-	}
-
 	// find the needed GitHub now
 	if op.ConnectionId == 0 {
 		return nil, errors.BadInput.New("connectionId is invalid")
 	}
+	if op.TransformationRules == nil && op.TransformationRuleId == 0 {
```
Comment (klesh) [inline plugins/github/tasks/task_data.go:63]: no more default value

Comment (klesh) [reply plugins/github/tasks/task_data.go:63]: https://github.com/apache/incubator-devlake/issues/3150

File: plugins/github/api/scope.go (hunk 5 above)
Comment (klesh) [reply plugins/github/api/scope.go:160]: Maybe we should rename `GithubId` to `ScopeId` instead?

File: plugins/github/api/scope.go (hunk 3 above)
Comment (mindlesscloud) [reply plugins/github/api/scope.go:154]: No, it's an unsigned integer

File: plugins/jira/models/board.go (hunk 1 above)
Comment (mindlesscloud) [reply plugins/jira/models/board.go:28]: I will remove `ScopeId` from tool layer tables

**Review by klesh — APPROVED** [review]

**Pull request merged**